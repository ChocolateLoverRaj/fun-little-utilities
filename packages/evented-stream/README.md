# Evented Stream

Collection of evented-stream packages

## License

[MIT](https://git.io/fxKXN) © [Hoàng Văn Khải](https://github.com/KSXGitHub)
