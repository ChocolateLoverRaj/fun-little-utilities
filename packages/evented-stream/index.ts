import * as types from 'evented-stream-types'
import iterateEventedStream from 'iterate-evented-stream'
import combineEventedStream from 'combine-evented-stream'

export * from 'evented-stream-types'
export { iterateEventedStream, combineEventedStream, types }

export const iterate = iterateEventedStream
export const combine = combineEventedStream
