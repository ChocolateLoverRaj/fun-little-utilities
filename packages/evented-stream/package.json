{
  "name": "evented-stream",
  "version": "0.1.0",
  "description": "Collection of evented-stream packages",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "keywords": [
    "stream",
    "iterate",
    "asynchronous",
    "event",
    "async",
    "iterable"
  ],
  "dependencies": {
    "tslib": "^1.10.0",
    "iterate-evented-stream": "^0.1.0",
    "combine-evented-stream": "^0.1.0",
    "evented-stream-types": "^0.1.0",
    "@types/node": "^12.7.8"
  },
  "engines": {
    "node": ">= 8.9.0"
  }
}
