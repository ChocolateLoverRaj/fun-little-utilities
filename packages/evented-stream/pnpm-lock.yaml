dependencies:
  '@types/node': 13.11.0
  combine-evented-stream: 'link:../combine-evented-stream'
  evented-stream-types: 'link:../evented-stream-types'
  iterate-evented-stream: 'link:../iterate-evented-stream'
  tslib: 1.11.1
lockfileVersion: 5.1
packages:
  /@types/node/13.11.0:
    dev: false
    resolution:
      integrity: sha512-uM4mnmsIIPK/yeO+42F2RQhGUIs39K2RFmugcJANppXe6J1nvH87PvzPZYpza7Xhhs8Yn9yIAVdLZ84z61+0xQ==
  /tslib/1.11.1:
    dev: false
    resolution:
      integrity: sha512-aZW88SY8kQbU7gpV19lN24LtXh/yD4ZZg6qieAJDDg+YBsJcSmLGK9QpnUjAKVG/xefmvJGd1WUmfpT/g6AJGA==
specifiers:
  '@types/node': ^13.11.0
  combine-evented-stream: ^0.1.7
  evented-stream-types: ^0.1.7
  iterate-evented-stream: ^0.1.8
  tslib: ^1.11.1
