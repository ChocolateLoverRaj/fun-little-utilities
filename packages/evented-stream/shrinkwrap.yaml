dependencies:
  '@types/node': 10.12.18
  combine-evented-stream: 'link:../combine-evented-stream'
  evented-stream-types: 'link:../evented-stream-types'
  iterate-evented-stream: 'link:../iterate-evented-stream'
  tslib: 1.9.3
packages:
  /@types/node/10.12.18:
    dev: false
    resolution:
      integrity: sha512-fh+pAqt4xRzPfqA6eh3Z2y6fyZavRIumvjhaCL753+TVkGKGhpPeyrJG2JftD0T9q4GF00KjefsQ+PQNDdWQaQ==
  /tslib/1.9.3:
    dev: false
    resolution:
      integrity: sha512-4krF8scpejhaOgqzBEcGM7yDIEfi0/8+8zDRZhNZZ2kjmHJ4hv3zCbQWxoJGz1iw5U0Jl0nma13xzHXcncMavQ==
registry: 'https://registry.npmjs.org/'
shrinkwrapMinorVersion: 9
shrinkwrapVersion: 3
specifiers:
  '@types/node': ^10.12.18
  combine-evented-stream: ^0.0.0
  evented-stream-types: ^0.0.0
  iterate-evented-stream: ^0.0.1
  tslib: ^1.9.3
