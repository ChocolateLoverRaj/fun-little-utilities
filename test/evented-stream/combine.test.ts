import { EventEmitter } from 'events'
import * as assets from 'monorepo-shared-assets'
import { iterateEventedStream, combineEventedStream, EventedStream } from 'evented-stream'
import getAsyncArray = assets.asyncIter.fns.getArray

enum StreamChunk {
  A,
  B,
  C,
  D,
}

class StreamError extends Error {
  constructor() {
    super('Expected')
    this.name = 'StreamError'
  }
}

abstract class MockedEmitter<Chunk extends StreamChunk = StreamChunk> extends EventEmitter
  implements EventedStream<Chunk, StreamError> {
  public emit(event: 'data', chunk: StreamChunk): boolean
  public emit(event: 'error', error: Error): boolean
  public emit(event: 'close'): boolean
  public emit(event: any, ...args: any[]): boolean {
    return super.emit(event, ...args)
  }
}

class FirstEmitter extends MockedEmitter<StreamChunk.A> {}
class SecondEmitter extends MockedEmitter<StreamChunk.B> {}
class ThirdEmitter extends MockedEmitter<StreamChunk.C> {}
class FourthEmitter extends MockedEmitter<StreamChunk.C> {}

type StreamCollection = [FirstEmitter, SecondEmitter, ThirdEmitter, FourthEmitter]

class Init {
  private timeout: number = 0

  private forward(fn: () => void, inc = 10): void {
    this.timeout += inc
    setTimeout(fn, this.timeout)
  }

  public readonly streamCollection: StreamCollection = [
    new FirstEmitter(),
    new SecondEmitter(),
    new ThirdEmitter(),
    new FourthEmitter(),
  ]

  public readonly combination = combineEventedStream<MockedEmitter, StreamChunk, StreamError>(this.streamCollection)

  public readonly emit = {
    data: <Chunk extends StreamChunk>(chunk: Chunk) =>
      this.forward(() => this.streamCollection[chunk].emit('data', chunk)),

    error: <Index extends StreamChunk>(index: Index) =>
      this.forward(() => this.streamCollection[index].emit('error', new StreamError())),

    close: () =>
      this.forward(
        () => this.streamCollection.forEach(stream => stream.emit('close')),
      ),
  }

  public readonly iterate = () => iterateEventedStream(this.combination)
  public readonly getArray = () => getAsyncArray(this.iterate())
  public readonly arrayPromise = this.getArray()
}

expect.addSnapshotSerializer({
  test(value: any): boolean {
    return value instanceof MockedEmitter
  },

  print(value: unknown): string {
    return `${(value as MockedEmitter<any>).constructor.name} {}`
  },
})

describe('when all streams emit data', () => {
  const { arrayPromise, emit, streamCollection } = new Init()
  const order: ReadonlyArray<StreamChunk> = [0, 3, 1, 2, 3, 0, 1, 1, 3, 0, 2, 3, 3, 3]
  order.forEach(emit.data)
  emit.close()

  it('matches snapshot', async () => {
    expect(await arrayPromise).toMatchSnapshot()
  })

  it('data have desired properties', async () => {
    expect(await arrayPromise).toEqual(
      order.map(chunk => ({ stream: streamCollection[chunk], chunk })),
    )
  })

  it('data have desired "stream" properties', async () => {
    ;(await arrayPromise).forEach(
      chunk => expect(chunk.stream).toBe(streamCollection[chunk.chunk]),
    )
  })

  it('data have desired valueOf() values', async () => {
    for (const chunk of await arrayPromise) {
      expect(chunk.valueOf()).toBe(chunk.chunk)
    }
  })
})

describe('when one of the streams emits error', () => {
  const { arrayPromise, emit, streamCollection } = new Init()
  const chosen = StreamChunk.C
  const order: ReadonlyArray<StreamChunk> = [0, 3, 1, 2, 3, 0, 1, 1, 3, 0, 2, 3, 3, 3]
  order.forEach(emit.data)
  emit.error(chosen)

  it('rejects with a reason that matches snapshot', async () => {
    await expect(arrayPromise).rejects.toMatchSnapshot()
  })

  it('rejects with a reason that has desired properties', async () => {
    await expect(arrayPromise).rejects.toMatchObject({
      stream: streamCollection[chosen],
      error: new StreamError(),
    })
  })

  it('rejects with a reason that has desired "stream" properties', async () => {
    const reason = await arrayPromise.then(
      () => {
        throw new Error('Expecting it to rejects but it resolves')
      },
      reason => reason,
    )

    expect(reason.stream).toBe(streamCollection[chosen])
  })

  it('rejects with a reason that has desired "error" properties', async () => {
    const reason = await arrayPromise.then(
      () => Promise.reject(new Error('Expecting it to reject, but it resolves')),
      reason => reason,
    )

    expect(reason.error).toBeInstanceOf(StreamError)
  })
})
