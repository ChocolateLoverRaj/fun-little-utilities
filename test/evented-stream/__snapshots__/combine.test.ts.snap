// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when all streams emit data matches snapshot 1`] = `
Array [
  ChunkCarrier {
    "chunk": 0,
    "stream": FirstEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 3,
    "stream": FourthEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 1,
    "stream": SecondEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 2,
    "stream": ThirdEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 3,
    "stream": FourthEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 0,
    "stream": FirstEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 1,
    "stream": SecondEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 1,
    "stream": SecondEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 3,
    "stream": FourthEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 0,
    "stream": FirstEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 2,
    "stream": ThirdEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 3,
    "stream": FourthEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 3,
    "stream": FourthEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
  ChunkCarrier {
    "chunk": 3,
    "stream": FourthEmitter {
      "_events": Object {},
      "_eventsCount": 0,
      "_maxListeners": undefined,
    },
  },
]
`;

exports[`when one of the streams emits error rejects with a reason that matches snapshot 1`] = `[ErrorCarrier: StreamError: Expected]`;
